### ISSUE TITLE
Tagging does not work if the sFlow Agent IP is an IPv6 address

### ISSUE THREAD

**Issue by toreanderson** [pmacct/pmacct#460]
**Description**
`sfacctd` appears to be unable to perform tagging according to the `pre_tag_map` if the incoming sFlow packets have an IPv6 Agent IP.

This minimal configuration reproduces the issue:

```console
[rlnc-user@pmacct pmacct]$ cat sfacctd.conf
aggregate: tag, proto
pre_tag_map: /etc/pmacct/pretag.map
[rlnc-user@pmacct pmacct]$ cat pretag.map 
set_tag=4       ip=0.0.0.0/0
set_tag=6       ip=::/0
```

This configuration *ought* have tagged all records with either `4` or `6` (depending on the IP version of the sFlow Agent IP). In other words, it should not be possible for any incoming records to miss both the `ip=0.0.0.0/0` and `ip=::/0` matches. However, that is *exactly* what happens if the incoming sFlow packets have an IPv6 Agent IP.

This is how the incoming stream of sFlow packets look like, according to `tshark` (anonymised):

```console
    1 0.000000000 2001:db8::HSFLOWD → 2001:db8::SFACCTD sFlow 1440 V5, agent 2001:db8::HSFLOWD, sub-agent ID 100000, seq 2357, 7 samples
    2 0.068651122 2001:db8::HSFLOWD → 2001:db8::SFACCTD sFlow 1448 V5, agent 2001:db8::HSFLOWD, sub-agent ID 100000, seq 2358, 8 samples
    3 0.115817995 2001:db8::HSFLOWD → 2001:db8::SFACCTD sFlow 1448 V5, agent 2001:db8::HSFLOWD, sub-agent ID 100000, seq 2359, 7 samples
```

Given this, I would have expected all records to be assigned a tag value of `6`. That does not happen:

```console
[rlnc-user@pmacct pmacct]$ pmacct -s
TAG         PROTOCOL    PACKETS               BYTES
0           icmp        12                    1124
0           esp         88                    27348
0           ipv6-icmp   5                     1915
0           udp         1957                  799471
0           tcp         16969                 12906874

For a total of: 5 entries
```
If, however, I change the `hsflowd` configuration to use an IPv4 Agent IP (without restarting or changing the configuration of `sfacctd`), everything starts working as expected:

```console
    1 0.000000000 2001:db8::HSFLOWD → 2001:db8::SFACCTD sFlow 1292 V5, agent 192.0.2.HSFLOWD, sub-agent ID 100000, seq 374, 8 samples
    2 0.146669420 2001:db8::HSFLOWD → 2001:db8::SFACCTD sFlow 1304 V5, agent 192.0.2.HSFLOWD, sub-agent ID 100000, seq 375, 7 samples
    3 0.215339977 2001:db8::HSFLOWD → 2001:db8::SFACCTD sFlow 1324 V5, agent 192.0.2.HSFLOWD, sub-agent ID 100000, seq 376, 7 samples
```

```console
[rlnc-user@pmacct pmacct]$ pmacct -e
OK: Clearing stats.
[rlnc-user@pmacct pmacct]$ pmacct -s
TAG         PROTOCOL    PACKETS               BYTES
4           ipv6-icmp   1                     120
4           tcp         2900                  2169611
4           icmp        3                     264
4           udp         347                   154666
4           esp         15                    4716

For a total of: 5 entries
```

Note for the bug to trigger the actual matches in the `pre_tag_map` does not matter. That is, it is not just matching on the sFlow Agent IP with `ip=` that fails, also matching on `filter=…` or anything else does. I am unable to tag based on *any* match while the sFlow Agent IP is an IPv6 address.

**Version**

I am using the official `sfacctd` container image from Docker Hub:

```console
[rlnc-user@pmacct pmacct]$ podman images
REPOSITORY                TAG     IMAGE ID      CREATED       SIZE
docker.io/pmacct/sfacctd  latest  b0fc1adadbb0  6 months ago  538 MB
[rlnc-user@pmacct pmacct]$ sfacctd -V

sFlow Accounting Daemon, sfacctd 1.7.5-git (20200510-00+c1)

Arguments:
 '--enable-mysql' '--enable-pgsql' '--enable-sqlite3' '--enable-kafka' '--enable-geoipv2' '--enable-jansson' '--enable-rabbitmq' '--enable-nflog' '--enable-ndpi' '--enable-zmq' '--enable-avro' '--enable-serdes' '--enable-redis' 'AVRO_CFLAGS=-I/usr/local/avro/include' 'AVRO_LIBS=-L/usr/local/avro/lib -lavro' '--enable-l2' '--enable-traffic-bins' '--enable-bgp-bins' '--enable-bmp-bins' '--enable-st-bins'

Libs:
libpcap version 1.8.1
MariaDB 10.3.22
PostgreSQL 110007
sqlite3 3.27.2
rabbimq-c 0.11.0-pre
rdkafka 1.5.0-pre1
jansson 2.13.1
MaxmindDB 1.4.2
ZeroMQ 4.3.2
Redis 0.15.0
avro-c
serdes
nDPI 3.2.0
netfilter_log

System:
Linux 4.18.0-193.14.2.el8_2.x86_64 #1 SMP Sun Jul 26 03:54:29 UTC 2020 x86_64

Compiler:
gcc 8.3.0

For suggestions, critics, bugs, contact me: Paolo Lucente <paolo@pmacct.net>.
```

**Appreciation**
> Please consider starring this project to boost our reach on github!

:point_down:
![image](https://user-images.githubusercontent.com/4557060/104091018-eff17a80-527a-11eb-89c1-8e7c731ce280.png)

**Comment by paololucente** [comment]
Hi @toreanderson ,

Thanks for getting in touch. Can you please send by unicast email even one single packet with the sFlow Agent set to an IPv6  address? That would allow me to easily reproduce the issue at my end and troubleshoot / fix the issue. From the sound of it, it looks like a bug.

Paolo

**Comment by toreanderson** [comment]
I just sent you an e-mail with a PCAP containing two sFlow packets, one with an IPv4 Agent IP, the other with an IPv6 Agent IP. Both are produced by `hsflowd` v2.0.11 running on Cumulus Linux v4.2, the only difference in the `hsflowd` configuration is the Agent IP setting.

Given the minimal `sfacctd` configuration described above, these two streams of sFlow yields the following:

```console
[rlnc-user@pmacct ~]$ pmacct -s
TAG         PROTOCOL    PACKETS               BYTES
0           icmp        20                    3541
4           ipv6-icmp   8                     1235
0           esp         171                   153316
0           ipv6-icmp   6                     888
0           udp         2690                  1154266
4           tcp         29172                 22721543
4           icmp        28                    3098
0           tcp         22957                 23655585
4           udp         2834                  1175078
4           esp         116                   28576

For a total of: 10 entries
```

The samples in the sFlow stream with an IPv4 Agent IP gets tagged correctly, the other does not.

**Comment by paololucente** [comment]
Thank you @toreanderson ,

This should be fixed now, see: https://github.com/pmacct/pmacct/commit/ce050a346951c8f123d2bddd97e15f28c3def130 . Can you give it a try?

Paolo

**Comment by toreanderson** [comment]
I tested `sfacctd 1.7.7-git (20210109-0 (ce050a3))` now (the current `bleeding-edge`-tag on Docker Hub), and there is a change of behaviour - tagging now happens, but it appears that the `ip=` match isn't _quite_ right. It appears that `ip=0.0.0.0/0` matches IPv6 agent IPs, and vice versa, `ip=::/0` matches IPv4 agent IPs.

That said, this is not a big deal for me. Those tags were just meant as a synthetic and minimal example to reproduce this issue, it is not representative of my intended use case. More important is that matching on exact IP adresses works as expected, and matching on other things (e.g., `filter='…'`) also works when the Agent IP is an IPv6 address. Thanks! :smiley:

**Comment by paololucente** [comment]
Right, so i am now handling 0.0.0.0/0 and ::/0 as equivalents. Before ::/0 would only match IPv6 sFlow agents whereas 0.0.0.0/0 would only match IPv4 sFlow agents -- and this all without being documented. Now both match both. Matching on exact IP addresses or IP prefixes is unchanged instead. Let me know if this helps and whether we can close this issue.

**Comment by toreanderson** [comment]
> Before ::/0 would only match IPv6 sFlow agents whereas 0.0.0.0/0 would only match IPv4 sFlow agents

That is IMHO the expected and intuitive behaviour.

However my (original) problem was that when the sFlow packets contained an IPv6 agent ID, *nothing* would match.

If the change you implemented in ce050a3 was only to consider `0.0.0.0/0` equal to `::/0` for the `ip=` match, then it might be possible that my original problem had already been fixed in master. I observed the problem in the latest release v1.7.5. As I didn't find any previously submitted bug reports about it I didn't re-test with master. I probably should have done so, now that I realise that v1.7.5 is actually quite old.

In any case, since matching/tagging now is possible, I am closing this issue. I might try bisecting v1.7.5..master later to determine whether or not the bug was already fixed (prior to ce050a3). Will update you with any findings.

**Comment by paololucente** [comment]
Please, yes. Keep me posted.

**Comment by toreanderson** [comment]
I have now bisected _v1.7.5..master_ and have determined that this issue had already fixed by 18c98919a4596a386dc0df9a30ac61119531f433. Closing, apologies for the noise.

As for the 0.0.0.0/0 vs ::/0 question, my expectation is that 0.0.0.0/0 means _"any IPv4 address"_ while ::/0 means _"any IPv6 address"_. I see you reverted ce050a3 now, so current _master_ behaves according to my expectations.

**Comment by paololucente** [comment]
Wonderful, thanks for confirming!
